### DISCUSSION TITLE
Permissions Based Tests for App

### DISCUSSION THREAD

**Discussion by ezrajrice** [tethysplatform/tethys#639] [Q&A]
I've been able to write tests for the Pull Requests I've made in the past and have been trying to update my apps with tests by following the guide and the sample that is generated with the creation of the app. I have been successful in some tests but not in others. Specifically there are two cases that are baffling me.

1. Testing URLs that are Permission/Permission Group limited.
    * I don't know how to create test users with specific permissions as defined in my application's app.py
2. Testing app URLs that use dynamic persistent stores.  
    * I have an app that creates a dynamic persistent store with a uuid4 name matching a uuid4 pk in the app's primary (static) persistent store. I need to know if I can create a test dynamic persistent store with predefined data to use for other tests. Is this even the right way of thinking about this sort of thing?

If anybody has done this successfully, I'd appreciate some help or pointers to get me going in the right direction. Thanks.

**Comment by swainn** [comment]
Hey @ezrajrice testing permissions is tricky, but hopefully, these tips will help. Here's the answer for part 1:

- You will need to use a `TethysTestCase` to be able to create users and assign them permissions in your tests (`from tethys_sdk.testing import TethysTestCase`).
- I use something like this [UserFactory](https://github.com/Aquaveo/tethysext-atcore/blob/master/tethysext/atcore/tests/factories/django_user.py) to create users for tests
- [Here is a test case that uses it](https://github.com/Aquaveo/tethysext-atcore/blob/fc65c97be9744b85820c233783f950a5ff237ccb/tethysext/atcore/tests/integrated_tests/controllers/app_users/resource_status.py#L22)
- At this point, you have a valid Django user and you can add it to any permission group like normal (including those you define in Tethys):

	```python
	from django.contrib.auth.models import Group
	group = Group.objects.get(name='group_name')
	group.user_set.add(user)
	```
- Assigning permissions to the user is a little more complicated b/c we use Django-Gaurdian to associate permissions with the Apps objects:

	```python
	from guardian.shortcuts import assign_perm
	from tethys_apps.models import TethysApp
	tethys_app = TethysApp.objects.get(package='my_first_app')
	assign_perm('permission_name', user, tethys_app)
	```
- The latter option is not recommended. Instead, assign all your Permissions to PermissionsGroups when you define them and then assign users to Groups.

**Reply by ezrajrice** [reply]
App permissions and permission groups don't show up in integration tests. It's as if the app isn't being initialized or loaded at all.

Running `print(list(Group.objects.all()))` in the setUp() method of a test class (inheriting from TethysTestCase) returns an empty list. When looking at my app, I have five permission groups added using Tethys's `PermissionGroup` class.

Likewise, running `print(list(Permission.objects.all()))` returns all of the Tethys permissions without any of the app permissions as defined in app.py

Lastly, when running an integration test on a method that uses the has_permission method provided by Tethys, the result is an error with the following traceback:
```bash
Traceback (most recent call last):
  File "/home/ezra/tethysdev/tethysapp-smartline_dlr/tethysapp/smartline_dlr/tests/integrated_tests/test_client_viewer/test_views.py", line 31, in test_cannot_access_organization_controller
    response = home(request)
  File "/home/ezra/tethys/miniconda/envs/tethys/lib/python3.7/site-packages/django/contrib/auth/decorators.py", line 21, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/home/ezra/tethysdev/tethysapp-smartline_dlr/tethysapp/smartline_dlr/controllers/organizations.py", line 40, in home
    can_view_records = has_permission(request, 'view_records')
  File "/home/ezra/tethys/miniconda/envs/tethys/lib/python3.7/site-packages/tethys_apps/base/permissions.py", line 131, in has_permission
    namespaced_perm = 'tethys_apps.' + app.package + ':' + perm
AttributeError: 'NoneType' object has no attribute 'package'
```

Digging into this error shows that the issue is actually in tethys_apps/utilities.py in the get_active_app method. There is an error being thrown where it is unable to get the app from the database and returns with the exception ObjectDoesNotExist 'Could not locate app with root url "smartline-dlr".' I had to manually run the check to this directly using the root_url defined in my app.py in order to figure out this was the error.

At this point I don't understand enough about the ins and outs of integration testing to make heads or tails of this. I was able to follow the "Tips and Tricks" discussion but that only helped me get the unit tests working. Any help would be appreciated.

**Reply by swainn** [reply]
@ezrajrice are you calling `super().setUp()` in the `setUp()` method where you added your `print` statements? If you just override it, I don't think the app will be loaded during the tests, and thus its permissions won't be loaded either. What does `print(list(TethysApp.objects.all()))` yield?

The `has_permission` error seems to indicate the apps are not loaded as well, hence app being `None`.

I might suggest a different strategy, which is to `mock.patch` the `has_permission` method:

```python
from unittest import mock

@mock.patch('path.to.my.controller.has_permission')
def test_my_test(self, mock_has_permission):
    mock_has_permission.return_value = False
    ...
```

**Reply by ezrajrice** [reply]
@swainn adding `super().setUp()` seems to have fixed the biggest issues. I still only have one of my five permission groups being added and recognized though. I'll try to jump on the next scrum to see if there's time to look at my issues.
